addons/stock/models/stock_traceability.py

```diff
@@ -0,0 +1,323 @@
+    if rec == 0:
+        rec = pStart
+    else:
+        rec += pInterval
+    return rec
+
+
+class MrpStockReport(models.TransientModel):
+    _name = 'stock.traceability.report'
+
+    @api.model
+    def get_move_lines_upstream(self, move_lines):
+        res = self.env['stock.pack.operation']
+        for move_line in move_lines:
+            # if MTO
+            if move_line.move_id.move_orig_ids:
+                res |= move_line.move_id.move_orig_ids.mapped('pack_operation_ids').filtered(
+                    lambda m: m.lot_id.id == move_line.lot_id.id)
+            # if MTS
+            else:
+                if move_line.location_id.usage == 'internal':
+                    res |= self.env['stock.pack.operation'].search([
+                        ('product_id', '=', move_line.product_id.id),
+                        ('lot_id', '=', move_line.lot_id.id),
+                        ('location_dest_id', '=', move_line.location_id.id),
+                    ])
+        if res:
+            res |= self.get_move_lines_upstream(res)
+        return res
+
+    @api.model
+    def get_move_lines_downstream(self, move_lines):
+        res = self.env['stock.pack.operation']
+        for move_line in move_lines:
+            # if MTO
+            if move_line.move_id.move_dest_ids:
+                res |= move_line.move_id.move_dest_ids.mapped('pack_operation_ids').filtered(
+                    lambda m: m.lot_id.id == move_line.lot_id.id)
+            # if MTS
+            else:
+                if move_line.location_dest_id.usage == 'internal':
+                    res |= self.env['stock.pack.operation'].search([
+                        ('product_id', '=', move_line.product_id.id),
+                        ('lot_id', '=', move_line.lot_id.id),
+                        ('location_id', '=', move_line.location_dest_id.id),
+                    ])
+        if res:
+            res |= self.get_move_lines_downstream(res)
+        return res
+
+    @api.model
+    def get_lines(self, line_id=None, **kw):
+        context = dict(self.env.context)
+        stream = context.get('ttype')
+        model = False
+        model_id = False
+        level = 1
+        parent_quant = False
+        if kw:
+            level = kw['level']
+            model = kw['model_name']
+            model_id = kw['model_id']
+            stream = kw['stream']
+            parent_quant = kw['parent_quant']
+        res = []
+        if context.get('active_id') and not context.get('model') or context.get('model') == 'stock.production.lot':
+            if stream == "downstream":
+                move_ids = self.env['stock.pack.operation'].search([
+                    ('lot_id', '=', context.get('active_id')),
+                    ('location_id.usage', '=', 'supplier'),
+                    ('state', '=', 'done'),
+                ])
+                res += self._lines(line_id, model_id=model_id, model='stock.pack.operation', level=level, parent_quant=parent_quant,
```

Comment:
this is coming from 46fc922 and we don't have time to clean it with this pr, please ignore diff from this commit :)